forte/data/extractor/base_extractor.py

```diff
@@ -0,0 +1,205 @@
+        """
+        return {
+            "vocab_method": "indexing",
+            "vocab_use_unk": True,
+            "need_pad": True,
+        }
+
+    @property
+    def state(self) -> Dict:
+        return {
+            "vocab_method": self.config.vocab_method,
+            "need_pad": self.config.need_pad,
+            "vocab_use_unk": self.config.vocab_use_unk,
+            "entry_type": self.config.entry_type,
+            "vocab": self.vocab.state if self.vocab else None,
+        }
+
+    @classmethod
+    def from_state(cls, state: Dict):
+        config = {
+            "vocab_method": state["vocab_method"],
+            "need_pad": state["need_pad"],
+            "vocab_use_unk": state["vocab_use_unk"],
+            "entry_type": state["entry_type"]
+        }
+        obj = cls(config)
+        if "vocab" in state and state["vocab"] is not None and \
+            state["vocab"] != "None":
+            obj.vocab = Vocabulary.from_state(state["vocab"])
+        else:
+            obj.vocab = None
+        return obj
+
+    @property
+    def entry_type(self) -> Type[Annotation]:
+        return self.config.entry_type
+
+    @property
+    def vocab_method(self) -> str:
+        return self.config.vocab_method
+
+    def get_pad_value(self) -> Union[None, int, List[int]]:
+        if self.vocab:
+            return self.vocab.get_pad_value()
+        else:
+            return None
+
+    def items(self) -> Iterable[Tuple[Hashable, int]]:
+        assert self.vocab
+        return self.vocab.items()
+
+    def size(self) -> int:
+        assert self.vocab
+        return len(self.vocab)
+
+    def add(self, element: Hashable):
+        assert self.vocab
+        return self.vocab.add_element(element)
+
+    def has_element(self, element: Hashable) -> bool:
+        assert self.vocab
+        return self.vocab.has_element(element)
+
+    def element2repr(self, element: Hashable) -> Union[int, List[int]]:
+        assert self.vocab
+        return self.vocab.element2repr(element)
+
+    def id2element(self, idx: int) -> Any:
+        assert self.vocab
+        return self.vocab.id2element(idx)
+
+    def get_dict(self) -> Dict[Hashable, int]:
+        assert self.vocab
+        return self.vocab.get_dict()
+
+    def predefined_vocab(self, predefined: Union[Set, List]):
+        """This function will add elements from the
+        passed-in predefined set to the vocab. Different
+        extractors might have different strategies to add
+        these elements. Override this function if necessary.
+
+        Args:
+            predefined (Union[Set, List]): A set or list contain
+                elements to be added into the vocabulary.
+        """
+        for element in predefined:
+            self.add(element)
+
+    def update_vocab(self, pack: DataPack,
+                    instance: Annotation):
+        """This function will extract the feature from
+        instance and add element in the feature to vocabulary.
+
```

Comment:
This function will also extract the feature?

Please document in a way that how users would implement this. The user will not be able to do so by looking this extractor class only.

If this function is supposed to update the vocab, I would suggest pass in the vocab object as parameter so the user can directly perform actions on it.

Reply:
Got. I will add construction on how to overwrite these function. I didn't think about that.
We have `predifined_vocab` for passed in vocabulary and `update_vocab` for building vocabulary incrementally.